File: backend-py/routers/instagram.py
"""
/api/instagram — post, reel, profile, thumbnail
"""

import asyncio
import os
from fastapi import APIRouter, HTTPException, BackgroundTasks
from fastapi.responses import FileResponse
from pydantic import BaseModel
from typing import Optional

from utils import clean_ig_url, extract_ig_username
from services.instagram_service import (
    download_post,
    download_reel,
    download_profile,
    download_ig_thumbnail,
)

router = APIRouter()


class PostRequest(BaseModel):
    url: str
    format: str = "mp4"
    start_time: Optional[str] = None
    end_time: Optional[str] = None


class ProfileRequest(BaseModel):
    # accepts either a username ("oiupoyt") or full URL ("instagram.com/oiupoyt/")
    username: str


class SimpleRequest(BaseModel):
    url: str


async def _delete_later(path: str, delay: int = 3600):
    await asyncio.sleep(delay)
    try:
        os.unlink(path)
    except Exception:
        pass


@router.post("/post")
async def post(req: PostRequest, bg: BackgroundTasks):
    url = clean_ig_url(req.url.strip())
    try:
        path, filename = await download_post(
            url=url, fmt=req.format,
            start_time=req.start_time, end_time=req.end_time,
        )
    except Exception as e:
        raise HTTPException(500, detail=str(e))
    bg.add_task(_delete_later, path)
    media_type = "video/mp4" if filename.endswith(".mp4") else "image/jpeg"
    return FileResponse(path, filename=filename, media_type=media_type)


@router.post("/reel")
async def reel(req: PostRequest, bg: BackgroundTasks):
    url = clean_ig_url(req.url.strip())
    try:
        path, filename = await download_reel(
            url=url, start_time=req.start_time, end_time=req.end_time,
        )
    except Exception as e:
        raise HTTPException(500, detail=str(e))
    bg.add_task(_delete_later, path)
    return FileResponse(path, filename=filename, media_type="video/mp4")


@router.post("/profile")
async def profile(req: ProfileRequest, bg: BackgroundTasks):
    """
    Accepts a username ("oiupoyt") or full URL ("https://www.instagram.com/oiupoyt/").
    Downloads the profile picture as a JPEG.
    """
    raw = req.username.strip()
    # if it looks like a URL, extract username from it
    if "instagram.com" in raw:
        try:
            username = extract_ig_username(raw)
        except ValueError as e:
            raise HTTPException(400, detail=str(e))
    else:
        username = raw.lstrip("@")

    if not username:
        raise HTTPException(400, detail="no username provided")

    try:
        path, filename = await download_profile(username)
    except Exception as e:
        raise HTTPException(500, detail=str(e))
    bg.add_task(_delete_later, path)
    return FileResponse(path, filename=filename, media_type="image/jpeg")


@router.post("/thumbnail")
async def thumbnail(req: SimpleRequest, bg: BackgroundTasks):
    url = clean_ig_url(req.url.strip())
    try:
        path, filename = await download_ig_thumbnail(url)
    except Exception as e:
        raise HTTPException(500, detail=str(e))
    bg.add_task(_delete_later, path)
    return FileResponse(path, filename=filename, media_type="image/jpeg")

File: backend/src/routes/instagram.rs
use axum::extract::State;
use axum::response::Response;
use axum::Json;
use serde::Deserialize;

use crate::config::AppConfig;
use crate::error::AppError;
use crate::services::instagram::InstagramService;
use crate::utils::{clean_ig_url, extract_ig_username};

#[derive(Deserialize)]
pub struct PostRequest {
    pub url: String,
    #[serde(default = "default_format")]
    pub format: String,
    pub start_time: Option<String>,
    pub end_time: Option<String>,
}

fn default_format() -> String {
    "mp4".to_string()
}

#[derive(Deserialize)]
pub struct ProfileRequest {
    pub username: String,
}

#[derive(Deserialize)]
pub struct SimpleRequest {
    pub url: String,
}

pub async fn post(
    State(config): State<AppConfig>,
    Json(req): Json<PostRequest>,
) -> Result<Response, AppError> {
    let _permit = config.semaphore.acquire().await.map_err(|_| {
        AppError::Internal("Server busy, please retry shortly".to_string())
    })?;

    let url = clean_ig_url(req.url.trim());
    let (path, filename) = InstagramService::download_post(
        &config.ytdlp_bin,
        &config.download_dir,
        &url,
        &req.format,
        req.start_time.as_deref(),
        req.end_time.as_deref(),
    )
    .await?;

    crate::routes::stream_file_response(path, filename).await
}

pub async fn reel(
    State(config): State<AppConfig>,
    Json(req): Json<PostRequest>,
) -> Result<Response, AppError> {
    let _permit = config.semaphore.acquire().await.map_err(|_| {
        AppError::Internal("Server busy, please retry shortly".to_string())
    })?;

    let url = clean_ig_url(req.url.trim());
    let (path, filename) = InstagramService::download_reel(
        &config.ytdlp_bin,
        &config.download_dir,
        &url,
        req.start_time.as_deref(),
        req.end_time.as_deref(),
    )
    .await?;

    crate::routes::stream_file_response(path, filename).await
}

pub async fn profile(
    State(config): State<AppConfig>,
    Json(req): Json<ProfileRequest>,
) -> Result<Response, AppError> {
    let _permit = config.semaphore.acquire().await.map_err(|_| {
        AppError::Internal("Server busy, please retry shortly".to_string())
    })?;

    let raw = req.username.trim();
    let username = if raw.contains("instagram.com") {
        extract_ig_username(raw).map_err(|e| AppError::BadRequest(e))?
    } else {
        raw.trim_start_matches('@').to_string()
    };

    if username.is_empty() {
        return Err(AppError::BadRequest("no username provided".to_string()));
    }

    let (path, filename) = InstagramService::download_profile(
        &config.ytdlp_bin,
        &config.download_dir,
        &username,
    )
    .await?;

    crate::routes::stream_file_response(path, filename).await
}

pub async fn thumbnail(
    State(config): State<AppConfig>,
    Json(req): Json<SimpleRequest>,
) -> Result<Response, AppError> {
    let _permit = config.semaphore.acquire().await.map_err(|_| {
        AppError::Internal("Server busy, please retry shortly".to_string())
    })?;

    let url = clean_ig_url(req.url.trim());
    let (path, filename) = InstagramService::download_thumbnail(
        &config.ytdlp_bin,
        &config.download_dir,
        &url,
    )
    .await?;

    crate::routes::stream_file_response(path, filename).await
}
